text/0000-panic-in-drop.md

```diff
@@ -0,0 +1,319 @@
+
+# Motivation
+[motivation]: #motivation
+
+There are two primary motivations for making this change: improving code size & compilation speed, and making it easier to write exception-safe unsafe code.
+
+## Exception safety
+
+Unsafe code must be [exception-safe](https://doc.rust-lang.org/nomicon/exception-safety.html). That is, it must ensure that it does not expose any potential unsoundness to safe code even when an unwinding panic occurs in the middle of it. This is particularly tricky to handle when the unsafe call must call user-provided functions that may panic, usually via trait methods. This is *exceptionally* tricky in the case of panics caused from dropping user-provided types since drops are usually invisible and implicit in the source code.
+
+Forgetting to take panics from drop into account has been the source of several bugs in popular crates and even the standard library:
+- Double drop in the standard library: https://github.com/rust-lang/rust/issues/83618
+- Double drop in `arrayvec`: https://github.com/bluss/arrayvec/issues/3
+- The exact same double-drop bug in `smallvec`: https://github.com/servo/rust-smallvec/issues/14
+
+### `catch_unwind`
+
+One particularly nasty case of missed exception safety is [rust-lang/rust#86027](https://github.com/rust-lang/rust/issues/86027).
+
+Code using `catch_unwind` is not typically prepared to handle an object that panics in its `Drop` impl. Even the standard library has had rious bugs in this regard, and if the standard library doesn't consistently get it right, we can hardly expect others to do so.
+
+`catch_unwind(code)` is often used to make sure no panics from `code` can cause further unwinding/panics. However, when catching a panic with a payload that panics on Drop, most usages of `catch_unwind(code)` will still result in further unwinding and often unsoundness.
+
+Example in the standard library (issue [rust-lang/rust#86030](https://github.com/rust-lang/rust/issues/86030)):
+
+```rust
+struct Bomb;
+
+impl Drop for Bomb {
+    fn drop(&mut self) {
+        panic!();
+    }
+}
+
+std::panic::panic_any(Bomb);
+```
+
+https://github.com/rust-lang/rust/blob/5ea19239d9d6f49fdd76513a36386d7e83708e3f/library/std/src/rt.rs#L34-L39
+
+```rust
+fn main() {
+    std::panic::panic_any(Bomb);
+}
+```
+```
+thread 'main' panicked at 'Box<Any>', src/main.rs:12:5
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread 'main' panicked at 'explicit panic', src/main.rs:7:9
+fatal runtime error: failed to initiate panic, error 5
+abort (core dumped)
+```
+
+### Non-obvious semantics of failing drops
+
+The exact semantics of a panicking drop are not obvious because it is often unclear what it means for a destructor to fail. The exact rules of panicking inside the `drop` function of an object seem to be:
```

Comment:
The "seem" here is a little confusing to me. Are these the semantics or not? What needs to happen to make this a stronger assertion?